include/pagmo/algorithms/gaco.hpp

```diff
@@ -0,0 +1,988 @@
+        std::vector<vector_double> sol_archive(m_ker, vector_double(1 + n_x + n_f, 1));
+        vector_double omega(m_ker);
+        vector_double prob_cumulative(m_ker);
+        std::uniform_real_distribution<> dist(0, 1);
+        std::normal_distribution<double> gauss{0., 1.};
+
+        // PREAMBLE-------------------------------------------------------------------------------------------------
+        // We start by checking that the problem is suitable for this
+        // particular algorithm.
+
+        if (!pop_size) {
+            pagmo_throw(std::invalid_argument, get_name() + " cannot work on an empty population");
+        }
+        if (prob.is_stochastic()) {
+            pagmo_throw(std::invalid_argument,
+                        "The problem appears to be stochastic " + get_name() + " cannot deal with it");
+        }
+        if (m_gen == 0u) {
+            return pop;
+        }
+        // I verify that the solution archive is smaller or equal than the population size
+        if (m_ker > pop_size) {
+            pagmo_throw(std::invalid_argument,
+                        get_name() + " cannot work with a solution archive bigger than the population size");
+        }
+        if (n_obj != 1u) {
+            pagmo_throw(std::invalid_argument, "Multiple objectives detected in " + prob.get_name() + " instance. "
+                                                   + get_name() + " cannot deal with them");
+        }
+
+        // ---------------------------------------------------------------------------------------------------------
+
+        // No throws, all valid: we clear the logs
+        m_log.clear();
+
+        // Main ACO loop over generations:
+        for (decltype(m_gen) gen = 1u; gen <= m_gen; ++gen) {
+            // At each generation we make a copy of the population into popold
+            population popold(pop);
+
+            auto dvs = pop.get_x(); // note that pop.get_x()[n][k] goes through the different individuals of the
+            // population (index n) and the number of variables (index k) the number of
+            // variables can be easily be deduced from counting the bounds.
+            auto fit = pop.get_f(); // The following returns a vector of vectors in which objectives, equality and
+            // inequality constraints are concatenated,for each individual
+
+            // I check whether the maximum number of function evaluations or improvements has been exceeded:
+            if (m_impstop != 0 && m_n_impstop >= m_impstop) {
+                std::cout << "max number of impstop exceeded" << std::endl;
+                return pop;
+            }
+
+            if (m_evalstop != 0 && m_n_evalstop >= m_evalstop) {
+                std::cout << "max number of evalstop exceeded" << std::endl;
+                return pop;
+            }
+
+            // 1 - compute penalty functions
+
+            // I declare some useful variables:
+            vector_double penalties(pop_size);
```

Comment:
Can you declare this variable outside the main loop, pre-allocating the necessary space? So that you allocate the vector once, instead of allocating it from scratch at every loop iteration.